packages/manager/src/features/Firewalls/FirewallDetail/Devices/AddLinodeDrawer.tsx

```diff
@@ -0,0 +1,233 @@
+          `Failed to add Linode ${label} (ID ${id}).`
+        )[0].reason;
+
+        if (!firstError) {
+          firstError = errorReason;
+        }
+
+        enqueueSnackbar(`Failed to add ${label}.`, { variant: 'error' });
+      }
+    });
+
+    setLocalError(firstError);
+    setSelectedLinodes(failedLinodes);
+
+    if (!firstError) {
+      onClose();
+    }
+  };
+
+  const errorNotice = () => {
+    let errorMsg = localError || '';
+    // match something like: Linode <linode_label> (ID <linode_id>)
+
+    const linode = /Linode (.+?) \(ID ([^\)]+)\)/i.exec(errorMsg);
+    const openTicket = errorMsg.match(/open a support ticket\./i);
+
+    if (openTicket) {
+      errorMsg = errorMsg.replace(/open a support ticket\./i, '');
+    }
+
+    if (linode) {
+      const [, label, id] = linode;
+
+      // Break the errorMsg into two parts: before and after the linode pattern
+      const startMsg = errorMsg.substring(
+        0,
+        errorMsg.indexOf(`Linode ${label}`)
+      );
+      const endMsg = errorMsg.substring(
+        errorMsg.indexOf(`(ID ${id})`) + `(ID ${id})`.length
+      );
+
+      return (
+        <Notice
+          sx={{
+            fontSize: '1rem',
+            fontWeight: 'bold',
+            lineHeight: '20px',
+          }}
+          variant="error"
+        >
+          {startMsg}
+          <Link to={`/linodes/${id}`}>{label}</Link>
+          {endMsg}
+          {openTicket ? (
+            <>
+              <SupportLink text="open a Support ticket" />.
+            </>
+          ) : null}
+        </Notice>
+      );
+    } else {
+      return <Notice text={localError} variant="error" />;
+    }
+  };
+
+  type OptionType = { id: number; label: string };
+
+  const currentLinodeIds =
+    currentDevices
+      ?.filter((device) => device.entity.type === 'linode')
+      .map((device) => device.entity.id) ?? [];
+
+  // If a user is restricted, they can not add a read-only Linode to a firewall.
+  const readOnlyLinodeIds = isRestrictedUser
+    ? getEntityIdsByPermission(grants, 'linode', 'read_only')
+    : [];
+
+  const optionsFilter = (linode: Linode) => {
+    return ![...currentLinodeIds, ...readOnlyLinodeIds].includes(linode.id);
+  };
+
+  const {
+    data,
+    error: linodeError,
+    isLoading: linodeIsLoading,
+  } = useAllLinodesQuery();
+
+  React.useEffect(() => {
+    if (linodeError) {
+      setLocalError('Could not load Linode Data');
+    }
+  }, [linodeError]);
+
+  const linodes = data?.filter(optionsFilter);
+
+  const options =
+    linodes?.map((linode) => ({ id: linode.id, label: linode.label })) || [];
+
+  const onChange = (selectedLinodes: OptionType[]) => {
+    const result = mapIdsToDevices<Linode>(
+      selectedLinodes.map((linode) => linode.id),
+      linodes
+    );
+
+    const mappedLinodes: Linode[] = Array.isArray(result)
+      ? result
+      : result
+      ? [result]
+      : [];
+
+    setSelectedLinodes(mappedLinodes);
+  };
```

Comment:
This function seems overly complex. We should be able to make it as simple as 

```suggestion
  const onChange = (selectedLinodes: Linode[]) => {
    setSelectedLinodes(selectedLinodes);
  };
```